src/Data/Array/Destination.hs

```diff
@@ -2,9 +2,95 @@
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
+-- | This module provides destination arrays
+--
+-- == What are destination arrays? What are they good for?
+--
+-- Destination arrays are write-only arrays that are only allocated once,
+-- thereby avoiding your reliance on GHC's fusion mechanisms to remove
+-- unneccessary allocations.
+--
+-- The current status-quo for computations that have a write-only array
+-- threaded along is to rely on fusion. While the optimizations in say,
+-- `Data.Vector` are quite good at ensuring GHC fuses, they aren't
+-- foolproof and can sometimes break by simple refactorings.
+--
+-- Avoiding extra allocations of a write-only array is easy in C, with
+-- something the functional programming world calls destination passing style,
+-- or DPS for short.
+--
+-- Here is a C function that manipulates an array written in DPS style; it
+-- takes in the destiniation array @res@ and writes to it:
+--
+-- @
+-- // ((a + b) * c) for vectors a,b and scalar c
+-- void apbxc(int size, int *a, int *b, int c, int *res){
+--   for (int i=0; i<size;++i){res[i]=a[i]+b[i];}
+--   mult(size, c, res);
+-- }
+--
+-- void mult(int size, int scalar, int* vec){
+--   for (int i=0; i<size; ++i){vec[i] *= scalar;}
+-- }
+-- @
+--
+-- We could write this with @Data.Vector@ in haskell and rely on GHC to fuse
+-- (but this isn't great in complex uses where some simple changes can
+-- break the fusion):
+--
+-- @
+-- import Data.Vector (Vector)
+-- import qualified Data.Vector as Vector
+-- import qualified Prelude
+--
+-- apbxc :: Vector Int -> Vector Int -> Int -> Vector Int
+-- apbxc a b c = mult c Prelude.$ Vector.zipWith (Prelude.+) a b
+--
+-- mult :: Int -> Vector Int -> Vector Int
+-- mult c = Vector.map (Prelude.* c)
+-- @
+--
+-- Doing this with destination arrays we are guarenteed only one allocation
+-- of the resulting array:
+--
+-- @
+--  apbxcDest :: Vector Int -> Vector Int -> Int -> Vector Int
+--  apbxcDest a b c = alloc (Vector.length a) writeResult
+--    where
+--      writeResult :: DArray Int #-> ()
+--      writeResult = fromFunction (\i -> ((a ! i) + (b ! i)) * c)
+-- @
+--
+-- == Aside: Why do we need linear types?
+--
+-- Linear types avoids ambigious writes to the destination array.
+-- For example, this function could never be linear and hence we avoid
+-- ambiguity:
+--
+-- @
+--  nonLinearUse :: DArray Int -> ()
+--  nonLinearUse arr = case (replicate 3 arr, replicate 4 arr) of
+--    ((),()) -> ()
+-- @
+--
+-- Further, linear types are used to ensure that each cell in the destination
+-- array is written to exactly once. This is because the only way to create and
+-- use a destination array is via
+--
+-- @
+-- alloc :: Int -> (DArray a #-> ()) #-> Vector a
+-- @
+--
+-- and the only way to really consume a @DArray@ is via our API
+-- which requires you to completely fill the array.
+--
```

Comment:
In here, we can also mention that they are performant since they are implemented with in-place updates on a mutable vector; which can be done safely because of linear-types.